lib/collection/src/shards/queue_proxy_shard.rs

```diff
@@ -120,11 +129,24 @@ impl QueueProxyShard {
     ///
     /// Because we have ownership (`self`) we have exclusive access to the internals. It guarantees
     /// that we will not process new operations on the shard while finalization is happening.
-    pub async fn finalize(self) -> Result<(LocalShard, RemoteShard), (CollectionError, Self)> {
+    ///
+    /// # Cancel safety
+    ///
+    /// This method is *not* cancel safe.
+    ///
+    /// If `cancel` is triggered - finalization may not actually complete in which case an error is
+    /// returned. None, some or all operations may be transmitted to the remote.
+    pub async fn finalize(
+        self,
+        cancel: CancellationToken,
+    ) -> Result<(LocalShard, RemoteShard), (CollectionError, Self)> {
         // Transfer all updates, do not unwrap on failure but return error with self
-        match self.transfer_all_missed_updates().await {
-            Ok(_) => Ok(self.forget_updates_and_finalize().await),
-            Err(err) => Err((err, self)),
+        match cancel_on_token(cancel, self.transfer_all_missed_updates()).await {
+            Ok(Ok(_)) => Ok(self.forget_updates_and_finalize().await),
+            // Transmission error
+            Ok(Err(err)) => Err((err, self)),
+            // Cancellation error
+            Err(err) => Err((err.into(), self)),
         }
     }
```

Comment:
This doesn't seem right. What's the difference if we explicitly cancel at `transfer_all_missed_updates` (e.g., with the `cancel_on_token`) or simply drop the whole `finalize` future? Either way, `forget_updates_and_finalize` may not be called.

IMO, we don't need to propagate `cancel` here, but I'm not yet sure if we should treat this as cancel-safe or not.

We call `revert_queue_proxy_local` explicitly if shard transfer task fails (which would call `forget_updates_and_finalize`), though.

Reply:
The issue is that `forget_updates_and_finalize()` is not cancel safe. That means we cannot simply drop this function.

The usual fix for this would be to spawn `forget_updates_and_finalize()` in a task (possibly through `spawn_cancel_on_drop`). But this doesn't seem possible due to ownership/lifetime issues.

Reply:
Disregard this.

`finalize` and `forget_updates_and_finalize` are not cancel safe, cause they consume `QueueProxyShard`.

But:
- we *have to* call `forget_updates_and_finalize` even if `transfer_all_missed_updates` returns an error
  - because `finalize` consumes `QueueProxyShard`
  - so there won't be `QueueProxyShard`
  - so we won't call `set_max_ack_version(None)` on error
    - *probably*, I didn't check the whole call stack yet
    - but it seems reasonable to handle it here anyway
- if `forget_updates_and_finalize` is handled this way, we can use `cancel_on_token` with `transfer_all_missed_updates`

Reply:
On second thought, dropping `forget_updates_and_finalize()` may actually be safe. The only issue is that we're not sure if we actually end up setting the max acknowledged version.

_This stuff is hard..._

Reply:
Yeah, it's *mostly* safe, with the only caveat of `set_max_ack_version`.

My original point was:
- if we *do* care about `set_max_ack_version`, then we *can't* simply use `cancel_on_token(cancel, self.transfer_all_missed_updates())`
  - because if we trigger `cancel`, we will exit before calling `set_max_ack_version`
  - and `finalize` consumes `self`, so it won't be available to the caller (to retry/explicitly handle `set_max_ack_version`) after we return `Err(Cancelled)`
- and if we *don't* care about `set_max_ack_version`
  - (e.g., we assume that it is not critical, or it should be handled explicitly somewhere else)
  - then we can simply mark `finalize` as cancel safe, and don't bother with `cancel` token

Reply:
Ok, I'm silly. I noticed that we return `self` on error. I guess it's fine as-is then. 🤦‍♀️

Reply:
> we _have to_ call `forget_updates_and_finalize` even if `transfer_all_missed_updates` returns an error

Not needed. Ownership is given back on error or on cancellation.

> if we do care about set_max_ack_version, then we can't simply use cancel_on_token(cancel, self.transfer_all_missed_updates())

Not necessarily. If the finalize call is cancelled, finalization will be incomplete. The user gets ownership back and has to finalize again.

---

So I believe the current implementation is fine. Or is it?

Reply:
@ffuugoo I think we concluded it's fine like this, so I'll resolve this comment. Please reopen it something is still outstanding.